central/node/datastore/store/postgres/store.go

```diff
@@ -488,22 +488,20 @@ func (s *storeImpl) isUpdated(ctx context.Context, node *storage.Node) (bool, bo
 		node.SetFixable = oldNode.GetSetFixable()
 		node.SetTopCvss = oldNode.GetSetTopCvss()
 	}
-	return true, scanUpdated, nil
+
+	return scanUpdated, nil
 }
 
 func (s *storeImpl) upsert(ctx context.Context, obj *storage.Node) error {
 	iTime := protoTypes.TimestampNow()
 
-	if !s.noUpdateTimestamps {
+	if !s.noUpdateTimestamps && obj.GetLastUpdated() == nil {
 		obj.LastUpdated = iTime
 	}
-	metadataUpdated, scanUpdated, err := s.isUpdated(ctx, obj)
+	scanUpdated, err := s.isUpdated(ctx, obj)
 	if err != nil {
 		return err
 	}
-	if !metadataUpdated && !scanUpdated {
-		return nil
-	}
```

Comment:
Help me understand this - why we should still do the upsert if there are no changes in both: the metadata and scan?

Reply:
Previously, `metadataUpdated` was always set to `true`, meaning this if the statement was redundant.

Reply:
I see. Shouldn't then we change 

```go
	if !metadataUpdated && !scanUpdated {
		return nil
	}
```

into

```go
	if !scanUpdated {
		return nil
	}
```
?

Reply:
Why?

Reply:
I see that the if-clause would never evaluate to true and we would never quit early here. However, I am thinking about the intentions of the original author of that code, and I think the following was meant: "if there were no changes to metadata and no changes to scan data, then quit early". By removing the entire if, we do keep the same behavior but we dismiss the original intention to quit early if there are no changes. Is there any value in quitting early `if !scanUpdated`?

Reply:
There is no value in quitting early `if !scanUpdate`. In fact, that is one of the cases we expect to occur when on `SensorEvent_Node`. The pipeline will update the Node with outdated (in fact `nil`) scan data. If we give up on that scenario, we prevent the update altogether. Unless there is a scenario in which both `last_updated` and `scan_time` are older, I am not aware of any.

Reply:
@md2119 scans can come now from two sources: nodes pipeline (aka V1 scan) & nodeInventory pipeline (aka V2 scan). The latter has richer set of components and vulnerabilities, but is reported every 4 hours. The Node pipeline has fewer scan data but is reported on every node update from Sensor. We do not want that I fresher V1 scan replaces older but still valid (and reacher) V2.

Is that what was meant here?

Reply:
If that is a guarantee, can we move the onus from the low-level store to a separate reconciler? Both node pipelines feed into this reconciler. When central receives node scans from either of the pipelines, the reconciler forwards or drops the message based on what you stated above. Therefore, the reconciler becomes the only channel/entry point from the datastore perspective. This way we move the complexity out of the low-level store.

Reply:
@jvdm what do you think about that idea?

Reply:
@md2119:

> I am not following the discussion entirely. Why would the an outdated scan have scan time later than recent scan time?

The `SensorEvent_Node` is sent when there are updates in the Kubernetes Node resource. That event used to be the only node event. And we would use the metadata (by that, I mean orchestrator, kernel, etc.) from the Node to generate scan components.

With Host-Level Node Scanning, we will not use the metadata from the `SensorEvent_Node` to generate scan components. Instead, with Host-Level Node Scanning, we have a new event, `SensorEvent_NodeInventory`, that arrives at every configurable interval (default ~4hs) with the scan components. We still need to handle `SensorEvent_Node`, but we don't need to populate the `Node.Scan` anymore when handling that event. That is only done when handling `SensorEvent_NodeInventory`.

That's the scenario where `Node.Scan` would be "outdated": When handling `SensorEvent_Node`. When writing the node to the store when handling `SensorEvent_Node`, the scan data should be ignored and pulled from the DB (if there is one). On the other hand, the other event `SensorEvent_NodeInventory` should not touch the Node metadata.